Fix negative expire values for refresh tokens and expire client session when user session

rmartinc opened the pull request:
Closes https://github.com/keycloak/keycloak/issues/14854
Closes https://github.com/keycloak/keycloak/issues/11990

I have joined both PRs into one. They are related and modify the same code area. Adding tests to check both. I have respected the original authors.

mposolda commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/util/SessionTimeouts.java:
```diff
@@ -218,23 +219,25 @@ public static long getOfflineClientSessionLifespanMs(RealmModel realm, Authentic
+        int userTimeToExpire = userSessionMaxLifespan - timeSinceUserSessionStart;
 
         // clientSession max lifespan has preference if set
+        int clientTimeToExpire = userTimeToExpire;
```
Same as above (but again, a nitpick...)

rmartinc replied:
Done!

mposolda commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/util/SessionTimeouts.java:
```diff
@@ -103,25 +103,26 @@ public static long getUserSessionMaxIdleMs(RealmModel realm, UserSessionEntity u
      * @return
      */
     public static long getClientSessionLifespanMs(RealmModel realm, AuthenticatedClientSessionEntity clientSessionEntity) {
```
Just thinking loud, that if this method also has `UserSessionModel` (or entity) as an argument, it will allow to:
- avoid introducing `USER_SESSION_STARTED_AT_NOTE` (which is another thing saved in the infinispan cache)
- allow more proper computation of `sessionMaxLifespan` with something like `int sessionMaxLifespan = userSession.isRememberMe() ? Math.max(realm.getSsoSessionMaxLifespan(), realm.getSsoSessionMaxLifespanRememberMe()) : realm.getSsoSessionMaxLifespan();`

On the other hand, it seems that having userSession available here may be fair amount of refactoring (and still not 100% sure if it is doable...). Also problem with not-100% accurate `sessionMaxLifespan` was even before this PR... So rather not worth an effort, but we're keeping as is now? Was just thinking if someone feels we should try to go this way, but I am myself rather not very keen on that...;-)

rmartinc replied:
Yeah! I also thought about getting the user session somehow. But I detected a big change (probably we would need the user session id in the authentication session) and I decided to follow contributors idea which seems to be easier.

mposolda replied:
yeah, agree. Thanks!

pjbober replied:
Hi, thanks @rmartinc for taking both changes in the fix. However, after I submitted my first PR I realized that to properly calculating the the lifespan we need to know if the user session has 'remember me' on or not, so apart from `USER_SESSION_STARTED_AT_NOTE` I added also note if it's remember me.
I got it @mposolda it's yet another thing to keep but in our specific case with remember me set to 180 days compared to couple hours without it, memory increase with these attribute is nothing with amount of sessions that are deleted thanks to proper calculation.

rmartinc replied:
@pjober If we also want to include the remember me then maybe it's better to directly store the user session id. I think it will need a big refactoring but it probably makes sense. @mposolda WDYT? (I suppose this will only be for the infinispan implementation.)

mposolda replied:
@rmartinc @pjbober Will it be possible to have "remember me" as another note on AuthenticatedClientSessionModel? The storing userSessionId instead on AuthenticatedClientSessionModel (instead of "Started at" and "remember me") works as well, but if it requires a big refactoring or additional lookup of userSession from infinispan, then my preference would be to rather stick with "Started at" and "remember me" notes.

mposolda reviewed (commented):
@rmartinc Thanks! In general, looks good.

I've added some very minor nitpick inline. Also I've added one question/concern inline, but just to receive some opinions on it (not really proposing we should refactoring this, but just thinking...)

@hmlnarik @mkanis I would like to also add store team to review this one. Also due the concern that this fix affects old store, and I am not sure if something similar is applicable for the new store as well (and should be added in this PR also to the new store?)

rmartinc commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/util/SessionTimeouts.java:
```diff
@@ -103,25 +103,26 @@ public static long getUserSessionMaxIdleMs(RealmModel realm, UserSessionEntity u
+        int userTimeToExpire = userSessionMaxLifespan - timeSinceUserSessionStart;
+
+        // clientSession max lifespan has preference but cannot be longer than the user session
+        int clientTimeToExpire = userTimeToExpire;
```
Just a little comment here to understand better the new calculation here. It fixes two different issues:

* The max life was using the timestamp and not the STARTED_AT of the client session, which is directly wrong. We were calculating wrong max life (it was more an idle timeout than a max life).
* Besides there is a second issue (non functional, just a memory issue) that client sessions can remain in memory more than the the user session. That's a waste of memory. Imagine a client session which is created just few seconds before than the user session expires, the user session will be removed but the client one (useless) remains in memory all the max-life of the client session. For this second problem the USER_SESSION_STARTED_AT_NOTE is needed.

Regards!

hmlnarik commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/entities/AuthenticatedClientSessionEntity.java:
```diff
@@ -42,6 +44,7 @@ public class AuthenticatedClientSessionEntity extends SessionEntity {
 
     // Metadata attribute, which contains the last timestamp available on remoteCache. Used in decide whether we need to write to remoteCache (DC) or not
     public static final String LAST_TIMESTAMP_REMOTE = "lstr";
+    public static final String USER_SESSION_STARTED_AT_NOTE = "userSessionStartedAt";
```
This needs to be in `AuthenticatedClientSessionModel`, not `AuthenticatedClientSessionEntity`

rmartinc replied:
OK! It's done now! I have also created the method `getUserSessionStarted` more or less the same to `getStarted`.

hmlnarik commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/RefreshTokenTest.java:
```diff
@@ -1044,37 +1049,195 @@ public void testUserSessionRefreshAndIdleRememberMe() throws Exception {
+        Integer clientIdleTimeout = realmResource.toRepresentation().getClientSessionIdleTimeout();
+        getTestingClient().testing().setTestingInfinispanTimeService();
+        try {
+            RealmManager.realm(realmResource).ssoSessionMaxLifespan(3600);
+            RealmManager.realm(realmResource).ssoSessionIdleTimeout(7200);
+            RealmManager.realm(realmResource).clientSessionMaxLifespan(5000);
+            RealmManager.realm(realmResource).clientSessionIdleTimeout(7200);
```
This effectively means four very slow calls to update realm parameters upon setting and another four upon resetting in the `finally` block. Please consider replacling all these calls in all methods affected in this PR with `RealmAttributeUpdater`, and removing the changes in `RealmManager`.

For an example usage, see https://github.com/keycloak/keycloak/blob/f5ebe675a845f26d338d7ccc64d8781fb6445597/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/LoginTest.java#L808-L812

rmartinc replied:
Yeah! Good point! Done in the added methods!

hmlnarik commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/entities/AuthenticatedClientSessionEntity.java:
```diff
@@ -83,6 +86,17 @@ public void setTimestamp(int timestamp) {
         this.timestamp = timestamp;
     }
 
+    public int getUserSessionStarted() {
+        String started = getNotes().get(USER_SESSION_STARTED_AT_NOTE);
+        // Fallback to timestamp if "started" note is not available.
+        return started == null ? timestamp : Integer.parseInt(started);
+    }
+
+    public int getStarted() {
+        String started = getNotes().get(AuthenticatedClientSessionModel.STARTED_AT_NOTE);
+        return started == null ? timestamp : Integer.parseInt(started);
+    }
+
```
These methods belong to either `AuthenticatedClientSessionModel` (not `AuthenticatedClientSessionEntity`) or into an utility method.

rmartinc replied:
I need to maintain both in the infinispan project. This class is not related at all to `AuthenticatedClientSessionModel` so I cannot reuse the methods from there. So creating them here is more or less the same than creating a utility class in the infinispan section.

Let me know if you have a better idea.

hmlnarik reviewed (changes requested):
Thank you @rmartinc .

Please see the comments inline. Please avoid modification of the entities and introducing new direct references from services module to legacy store modules (infinispan in this case). These changes tightly couple services with legacy store which complicates the aim to decouple the two.

martin-kanis reviewed (changes requested):
@rmartinc Thanks for the PR. I suspect these problems might be present in the new store as well. Currently, `OfflineTokenTest` and `RefreshTokenTest` classes are not run with the new store in GHA. When I tried to enable them in [GHA](https://github.com/martin-kanis/keycloak/actions/runs/4404125754/jobs/7713951995), your tests failed with 500 because the tests rely on legacy store specifics (i.e `testingClient.testing().cache(InfinispanConnectionProvider.OFFLINE_USER_SESSION_CACHE_NAME).contains(sessionId));`)
Would it be possible to use just methods available in `UserSessionProvider`, `UserSessionModel` and `AuthenticatedClientSessionModel`?
PS: Some failures with the new storage in existing tests might not be cased by this PR. Michal is working on it https://github.com/keycloak/keycloak/pull/17572

rmartinc commented:
> Would it be possible to use just methods available in UserSessionProvider, UserSessionModel and AuthenticatedClientSessionModel?

Not completely, the the user-session and the client-session are separated in infinispan. It's possible that the user-session disappears (by max-life for example) but the client-session remains. But via API retrieving a client session without the user-session is not possible (because it just makes sense in infinispan). So I cannot assert that the client session is fully removed. At least I have to maintain the last part [here](https://github.com/keycloak/keycloak/blob/8f9d327bda798294c1afed9aca0de5f8309a75ff/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/RefreshTokenTest.java#L1104).

I can change the other checks but not this one (I need access to the real infinispan API for that). I can also not assert this part but it was exactly one of the error I was trying to fix in the PR.

rmartinc commented:
@martin-kanis I have replaced all the uses that can be replaced. And for the one that cannot be substituted I did [this trick](https://github.com/keycloak/keycloak/blob/94ec7a37dd0e7c7cf30d874c92e7ccaf1c800a07/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/RefreshTokenTest.java#L1118-L1120). Let me know if it works for you. (I really want to test that the session is really gone from the infinispan cache.)

martin-kanis commented:
> @martin-kanis I have replaced all the uses that can be replaced. And for the one that cannot be substituted I did [this trick](https://github.com/keycloak/keycloak/blob/94ec7a37dd0e7c7cf30d874c92e7ccaf1c800a07/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/RefreshTokenTest.java#L1118-L1120). Let me know if it works for you. (I really want to test that the session is really gone from the infinispan cache.)

Thank you. I run again this PR with new store and there are these failures:

OfflineTokenTest.refreshTokenUserClientMaxLifespanSmallerThanSession:1036 expected:<1> but was:<0>
RefreshTokenTest.refreshTokenUserClientMaxLifespanSmallerThanSession:1166 expected:<1> but was:<0>
This is probably caused by the difference how lifespans and expirations are computed for the new store.

It would be good to have the other tests you added running fully with the new store. Either by transferring them to model tests as Hynek suggested, or at least checking if `hotrod` map storage is running and then looking directly into Infinispan caches. Let me know if you need an assistance with any of those.

rmartinc commented:
@martin-kanis and all, I have been doing some tests and created a model test class ([this one in a test branch](https://github.com/rmartinc/keycloak/blob/TEST/testsuite/model/src/test/java/org/keycloak/testsuite/model/infinispan/SessionTimeoutsTest.java)). I can say that there are differences between infinispan and the new map for session lifespans. The hot-rod fails in general (don't know why).

So I think that first we should agree how this should work. There are two main differences:

- In infinispan the user session does not take into account the realm client session values. The calculation is [here](https://github.com/keycloak/keycloak/blob/21.0.1/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/util/SessionTimeouts.java#L51-L67). But the map expiration does take into account client values, [here](https://github.com/keycloak/keycloak/blob/21.0.1/model/map/src/main/java/org/keycloak/models/map/userSession/SessionExpiration.java#L138-L141). So in the new map impl the user session lifespan is the min time between "SSO Session Max" and "Client Session Max" but in infinispan "Client Session Max" is only for client session (never for the user session). IMHO infinispan interpretation is the correct one in this point.
- The second difference is the client lifespan in infinispan does not use the ClientModel to check the attribute for overrides for the client session lifespan (only for tokens). In map it is taken into account [here](https://github.com/keycloak/keycloak/blob/21.0.1/model/map/src/main/java/org/keycloak/models/map/userSession/SessionExpiration.java#L73-L77). IMHO the new map is doing the right thing here although changing this in infinispan will be complicated.

Let's wait for @mposolda too, but I think that we have to decide how to handle those two points and do the same for both impls.

rmartinc committed:
Fix and correlate session timeout calculations in legacy and new map implementations
Closes https://github.com/keycloak/keycloak/issues/14854
Closes https://github.com/keycloak/keycloak/issues/11990

rmartinc commented:
@mposolda @hmlnarik @martin-kanis The previous week I started to work again on this. I think that everything is in place and working. The change is quite big now, as I moved timeout calculations to a shared utility class and the two implementations perform the same timeouts:

* Client session times (idle or max) never affect the user session timeouts. In this point I'm following what the old impl was doing. IMHO it makes no sense doing anything different, because setting a different value in "SSO Session Max" and "Client Session Max" would be worthless. And it can make sense that someone wants a big user session but short client sessions for all their apps because they are very long lived.
* The attribute in the clients for client session max/idle are always used for the expiration of the session. This is what the new impl was doing and a typical complain for the old impl (for example #15829). This point is the one that introduces most of the changes of the PR (adding it for old infinispan was not easy).
* The "SSO Session Idle/Max Remember Me" and "SSO Session Idle/Max" are used like in the old impl, taking the max value between both settings for online sessions (it never affected offline). The new impl chose the first value for remember me sessions.

In general we can decide something different in any point. I just decided what it's more logical IMHO and it was less risky.

The PR has been rebased. Added different tests in different places. And there are some minor tests fixes because now calculations are different for both impls (this affected two previous tests).

martin-kanis reviewed (changes requested):
@rmartinc Thank you for the update. I noticed this PR brings some instability into model tests (`*cross-dc*`) profiles. See my branch where I run model tests 10 times https://github.com/martin-kanis/keycloak/actions/runs/5119182601/jobs/9204403790
Also please see my other comments:

mposolda commented on server-spi-private/src/main/java/org/keycloak/models/utils/SessionExpirationUtils.java:
```diff
@@ -0,0 +1,203 @@
+            long clientSessionCreated, long userSessionCreated, RealmModel realm, ClientModel client) {
+        long timestamp = -1;
+        if (offline) {
+            if (realm.isOfflineSessionMaxLifespanEnabled()) {
```
Let's assume that:
- Realm settings have "Offline Session Max Lifespand" disabled
- Client has `Client Offline Token Max` overriden to 1 day.

IMO the correct offline client session timeout should be 1 day with those settings? However it looks that current logic in this PR will just use `never` (or `-1`) with those settings due this "if" check, which causes that disabled `Offline Session Max Lifespan` won't take into account overriden values for `Client Offline Token Max` ?

rmartinc replied:
I prefer to maintain this for the moment. It was like this in both implementations. In order to have expiration for offline session the realm `isOfflineSessionMaxLifespanEnabled` should be enabled first. If not offline sessions never expire by lifespan.

mposolda replied:
Yes, that works for me.

However it will be good to improve the usability around this IMO, but that can be done as a follow-up. Currently if I configure the things in the admin console as I've mentioned above (`Offline Session Max Lifespan` disabled for the realm, but `Client Offline Token Max` overriden to 1 day for client `foo`), I would expect that offline client session for my client `foo` would expire in 1 day.

The possibility here is maybe to completely hide the `Client Offline Token Max` switch for the clients in case that realm `Offline Session Max Lifespan` is disabled. As the `Client Offline Token Max` client switch in-fact doesn't have any effect in this case and hence exposing it in the admin console is confusing IMO. But I see that as a possible follow-up...

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/RefreshTokenTest.java:
```diff
@@ -1045,37 +1050,185 @@ public void testUserSessionRefreshAndIdleRememberMe() throws Exception {
+                    r.setSsoSessionMaxLifespan(3600);
+                    r.setSsoSessionIdleTimeout(7200);
+                }).update()) {
+            RealmManager.realm(realmResource).ssoSessionMaxLifespan(3600);
```
Minor: lines 1085 and 1086 seems to be redundant due the timeout is set in the updater above?

rmartinc replied:
done!

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/RefreshTokenTest.java:
```diff
@@ -1132,6 +1285,79 @@ public void refreshTokenUserSessionMaxLifespanWithRememberMe() throws Exception
+        getTestingClient().testing().setTestingInfinispanTimeService();
+
+        try {
+            rep.setSsoSessionMaxLifespan(10);
```
Is it possible to use bigger values for this test? For example 100 instead of 10 and also multiply all subsequent calls to `setTimeOffset` by 10? AFAIR in some slow environments, small times/timeout values can cause headaches and test instabilities. As in some environments, this test can easily take 3 seconds or more, which could result in the test failure.

rmartinc replied:
Yes, no problem, this test was taken from one of the previous PRs. I have used *100. So now the times are 1000, 500, ans so on and so forth.

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/OfflineTokenTest.java:
```diff
@@ -829,10 +837,35 @@ private void conductOfflineTokenRequest(String expectedRefreshAlg, String expect
+        }, String.class);
+    }
+
+    private int countSessions(final String userSessionId, final String clientId, final String clientSessionId) {
```
Minor: countSessions is likely a bit misleading name of this method as it doesn't return any "count" of sessions? Could be maybe renamed to something like `getTestSessionsState` (or something even better if you can think of something)?

rmartinc replied:
In my defense I'm really counting sessions :smile:. 0 is returned if no user and client session exists, 1 if only user session is present, and 2 if both are found. The client session alone cannot be detected because it needs the user session to be retrieved. I changed the method to `checkIfUserAndClientSessionExist`.

mposolda reviewed (changes requested):
@rmartinc Thanks for the PR update! I think it makes sense that "old" and "new" store are unified in the behaviour and the timeout changes from this PR makes sense to me.

I've added few comments inline. 

Besides that, it will be nice to clarify tooltips and docs IMO. I suppose that it's hard to add much info to the tooltips, but maybe this chapter https://www.keycloak.org/docs/latest/server_admin/index.html#_timeouts can have some more details added? There is nothing about client timeouts right now. The fact that it's possible to override timeouts at the client level would be nice addition IMO and would makes sense to clear-ify that client overriden timeouts have preference over the realm timeouts. Also possibly some warning that lower value will have precedence (EG. in case that client timeouts are bigger than user session timeouts, then it would effectively use user session timeouts etc). Also not sure if "Upgrading guide" should have something, but not 100% sure, as this is pretty much fixing the current behaviour? I am not sure if docs should be added to this PR or to the follow-up PR? To me, follow-up PR is completely fine (and maybe even better option). WDYT?

Anyway, before updating docs (and tooltips), I suggest to wait for the review/approval from the store team for this PR.

Other possible follow-ups. I suggest to wait for the feedback to this PR from store team before starting on them:

- It seems to me that methods `TokenManager.getRefreshExpiration` and `TokenManager.getOfflineExpiration()` have some duplicated logic for retrieving timeouts. Maybe it's possible to adjust those methods to re-use the logic from `SessionExpiration` class to have unified place where the logic for timeouts is implemented? But I suggest to do this refactoring possibly as a follow-up (once we agree with store team that this is a way to go).


- When I open client in the admin console -> then tab "Advanced" -> I can see that `Client Offline Token Max` is set to `Inherits from realm settings` -> `60 days`. However this seems to be slightly misleading as at the realm level, I have `Offline Session Max Limited` disabled. Which means that `Client Offline Token Max` is also in fact disabled AFAIK? It might be probably better that if `Offline Session Max Limited` is disabled at realm level, then client will show something like `Disabled` or `Never expire`. This might be more a task for UI team rather than core team IMO (TBD).

@rmartinc @martin-kanis @hmlnarik WDYT?

rmartinc commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/SessionTimeoutsTest.java:
```diff
@@ -0,0 +1,435 @@
+ */
+package org.keycloak.testsuite.model.session;
+
+import java.util.Arrays;
```
Fixed!

rmartinc commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/SessionTimeoutsTest.java:
```diff
@@ -0,0 +1,435 @@
+        });
+
+        try {
+            final String[] sessions = inComittedTransaction(session -> {
```
I have changed to use `withRealm` always, I always need the realm. Thanks!

rmartinc commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/SessionTimeoutsTest.java:
```diff
@@ -0,0 +1,435 @@
+        }
+    }
+
+    protected void testUserClientIdleTimeoutSmallerThanSession(int refreashTimes, boolean offline, boolean overrideInClient) {
```
Upps! Fixed!

rmartinc commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/SessionTimeoutsTest.java:
```diff
@@ -0,0 +1,435 @@
+
+            // set client session max lifespan smaller at realm or client
+            ClientModel client = realm.getClientByClientId("test-app");
+            if (overrideInClient) {
```
Ok, I have created a method called `configureTimeouts` that is called to set the timeouts in the three tests.

rmartinc commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/SessionTimeoutsTest.java:
```diff
@@ -0,0 +1,435 @@
+
+    protected void testUserClientMaxLifespanSmallerThanSession(boolean offline, boolean overrideInClient) {
+        withRealm(realmId, (session, realm) -> {
+            realm.setOfflineSessionMaxLifespanEnabled(true);
```
New method `configureTimeouts` is used now for all the tests.

rmartinc commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/SessionTimeoutsTest.java:
```diff
@@ -0,0 +1,435 @@
+            setTimeOffset(1000);
+
+            inComittedTransaction(session -> {
+                // refresh sessions at half-time => both session should exist
```
Yes, I did so many changes in the tests that the comment was meaningless now. Thanks!

rmartinc commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/SessionTimeoutsTest.java:
```diff
@@ -0,0 +1,435 @@
+            inComittedTransaction(session -> {
+                // ensure user session is expired after user session expiration
+                RealmModel realm = session.realms().getRealm(realmId);
+                InfinispanConnectionProvider prov = session.getProvider(InfinispanConnectionProvider.class);
```
Yep, I was checking with infinispan first, but then I just searched sessions to work in all the backends. This was a leftover I missed. Thanks!

rmartinc commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/SessionTimeoutsTest.java:
```diff
@@ -0,0 +1,435 @@
+                : session.sessions().getUserSession(realm, id);
+    }
+
+    protected static String cacheName(boolean user, boolean offline) {
```
The same, a leftover used when I was testing with infinispan. I missed to removed it. Thanks!

rmartinc commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/SessionTimeoutsTest.java:
```diff
@@ -0,0 +1,435 @@
+        clientModel.setSecret("password");
+    }
+
+    protected static RealmModel createRealm(KeycloakSession s, String name) {
```
Done!

rmartinc commented on testsuite/model/src/test/java/org/keycloak/testsuite/model/session/SessionTimeoutsTest.java:
```diff
@@ -0,0 +1,435 @@
+
+    protected static void createClients(KeycloakSession s, RealmModel realm) {
+        ClientModel clientModel = s.clients().addClient(realm, "test-app");
+        clientModel.setEnabled(true);
```
Right! I just copied this from another test. :smile: Its done now!

rmartinc commented:
@martin-kanis @mposolda I think that I have done all the modifications. I also investigated the inestability. It seems that the new class can be flaky in cross-dc environments. Checking the issue it seems that sometimes the session is found although it should be expired (for example in this [line](https://github.com/keycloak/keycloak/blob/510cc404c531b9faa4f7dacc5817e6e7d38971ef/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/SessionTimeoutsTest.java#L283)). But I have checked that this was already happening for example in this flaky issue #16511. That issue is the same situation, sometimes in cross-dc this line [line](https://github.com/keycloak/keycloak/blob/21.1.1/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/UserSessionProviderOfflineModelTest.java#L208) returns the session although it should be expired. So there is something wrong in cross-dc.

The big problem was that after that test failed (although it run OK in the 2nd run, just flaky). Another test always failed because the session of the failed tests remained in the persistent store. I have fixed that part removing sessions in the `clearEnvironment` method ([here](https://github.com/keycloak/keycloak/blob/510cc404c531b9faa4f7dacc5817e6e7d38971ef/testsuite/model/src/test/java/org/keycloak/testsuite/model/session/SessionTimeoutsTest.java#L75)). So now the other tests are not affected by the flaky run.

I don't want to fix the cross-dc issue in this PR. I'm just changing the timeouts calculations here and that error seems to be happening before. :stuck_out_tongue_winking_eye:

@martin-kanis Run the tests again 10 times to see what happens now.

rmartinc commented:
Regarding @mposolda comments, I think that the configuration is quite convoluted in general. But I'm just trying to fix the current issues and execute the same code for all the implementations. After that we can improve it more, but, for the moment, I just want to use the same methods. I tried to do the minimal changes and do something that makes sense.

I agree that if this gets merged, we probably should modify documentation and tooltips again. We are changing how this works, not much, but we are changing it.

mposolda reviewed (dismissed):
I am approving as from my point of view, the changes look good! I consider that we have smaller follow-up tasks, which we can/should do as a follow-up PRs (documentation, tooltips, new admin console polishing, possible TokenManager refactoring to leverage methods from SessionExpiration).

I didn't much review parts related to new store and model tests, as leaving this to "store" team.

martin-kanis commented:
@rmartinc The updated version in now green after 10 runs https://github.com/martin-kanis/keycloak/actions/runs/5130604990/jobs/9229608348

rmartinc commented:
> @rmartinc The updated version in now green after 10 runs https://github.com/martin-kanis/keycloak/actions/runs/5130604990/jobs/9229608348

:heart_eyes: Great! But if you see `SessionTimeoutsTest#testOfflineUserClientIdleTimeoutSmallerThanSessionNoRefresh` is flaky. The reason is in my previous comment. Maybe checking idle so rigorously is not OK, I can improve the test to do N checks inside, and only fail if the idle is wrong after the N times. WDYT?

rmartinc commented:
OK, I have tested different things with the CI (as I cannot reproduce the issue in my laptop) with the profile `legacy-jpa+cross-dc-infinispan-offline-sessions-preloading` and trying to execute a single test 100 times:

* testOfflineUserClientIdleTimeoutSmallerThanSessionNoRefresh: failed at 4th (check idle without any refresh).
* testOfflineUserClientIdleTimeoutSmallerThanSessionOneRefresh: failed at 31st (check idle with one refresh).
* testOfflineUserClientMaxLifespanSmallerThanSession: failed at 54th run (check max lifespan).

So, it's not just idle, it can happen in max lifespan too. It seems to be more probable with idle, but it also happens with max lifespan. So I'm not doing anything more, I will wait for @martin-kanis comments. As commented the only thing I see to improve this at keycloak side is always checking expiration before returning the sessions (or at least in some scenarios like cross-dc).

Cheers!

martin-kanis commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProvider.java:
```diff
@@ -287,7 +290,12 @@ private UserSessionEntity getUserSessionEntityFromPersistenceProvider(RealmModel
-        return importUserSession(realm, offline, persistentUserSession);
+        UserSessionEntity sessionEntity = importUserSession(realm, offline, persistentUserSession);
+        if (sessionEntity == null) {
+            persister.removeUserSession(sessionId, offline);
```
Out of curiosity, why we need to delete the offline session if `importUserSession` didn't find in the DB? Is it because we want to remove associated offline client sessions?

rmartinc replied:
At this point we now the session is in the offline store but is expired. So it's removed now. If not it will be removed by some scheduled task later. I think that I needed this for the tests (because the sessions created by the new test remained in the store and other tests related with the persister counted them).

martin-kanis commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/entities/AuthenticatedClientSessionEntity.java:
```diff
@@ -83,6 +85,28 @@ public void setTimestamp(int timestamp) {
 
+    public int getUserSessionStarted() {
+        String started = getNotes().get(AuthenticatedClientSessionModel.USER_SESSION_STARTED_AT_NOTE);
+        return started == null ? timestamp : Integer.parseInt(started);
```
This defaults to `timestamp` while `AuthenticatedClientSessionModel.getUserSessionStarted` defaults to 0 if `USER_SESSION_STARTED_AT_NOTE` is not set. Is it intentional?

rmartinc replied:
Yes, it's correct. Previously `getTimestamp` was always used. So this is maintaining the same behavior just in case. If we use 0 all the session without the note would be expired. Note that we are doing the same in the map part ([here](https://github.com/keycloak/keycloak/blob/510cc404c531b9faa4f7dacc5817e6e7d38971ef/model/map/src/main/java/org/keycloak/models/map/userSession/SessionExpiration.java#L32)).

martin-kanis replied:
But why we don't use `timestamp` in `AuthenticatedClientSessionModel` but 0 instead?

```
default int getUserSessionStarted() {
        String started = getNote(USER_SESSION_STARTED_AT_NOTE);
        return started == null ? 0 : Integer.parseInt(started);
}
```
and older code

```
default int getStarted() {
        String started = getNote(STARTED_AT_NOTE);
        // Fallback to 0 if "started" note is not available. This can happen for the offline sessions migrated from old version where "startedAt" note was not yet available
        return started == null ? 0 : Integer.parseInt(started);
    }
```

rmartinc replied:
We can do it, but I really don't know if it can be problematic. In `AuthenticatedClientSessionEntity` I simply created `getUserSessionStarted` following the same idea that was in `getStarted` (so I returned 0).

For the PR the classes that are really used are `AuthenticatedClientSessionEntity` and `MapAuthenticatedClientSessionEntity` (which don't implement `AuthenticatedClientSessionEntity` and re-implement the access to notes).

martin-kanis replied:
@mposolda Do you recall why you used 0 as default? To me, this implies that `isIssuedBeforeSessionStart` check will be always `false` meaning "legacy" sessions will always pass this check. At the same time, this PR adds https://github.com/keycloak/keycloak/pull/17525/files#diff-fd8b1c47da5ec7ee42d5450d3f7b73cdeef9c08af26b825c9245e3e6052c5cdfR1090 and https://github.com/keycloak/keycloak/pull/17525/files#diff-fd8b1c47da5ec7ee42d5450d3f7b73cdeef9c08af26b825c9245e3e6052c5cdfR1130 that uses `timestamp` instead 0.

Where I am heading with this, isn't better to use `timestamp` at the first place unless Marek remembers reasons why he used 0?

rmartinc replied:
Yes, @martin-kanis, I can give it a try to see it brings any issue. We can change both methods to return the timestamp instead of 0 in `AuthenticatedClientSessionEntity`.

mposolda replied:
@martin-kanis @rmartinc This is good catch!

I recall this was added in this commit https://github.com/keycloak/keycloak/pull/8130/files#diff-b4014a5e9c73d54dc98153e6fa64803e21ab0646a75566cf74ca9a02abbb18db . And some related story behind that is described in this JIRA description https://issues.redhat.com/browse/KEYCLOAK-18368 .

The 0 was just a fallback, which was only used for "offline client sessions" migrated from previous version (As those offline client sessions didn't have "startedAt" note on them). In other words when it would use timestamp instead of 0 in that PR, we wouldn't be able to refresh offline tokens from previous version due this possible scenario:
- Offline token issued on old server (before KEYCLOAK-18368) at time 1000
- Keycloak server stopped and started with new Keycloak version (with KEYCLOAK-18368 applied)
- Refresh of the offline token would fail due this check https://github.com/keycloak/keycloak/blob/21.1.1/services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java#L186-L189

Note that this check is needed exactly because of the scenario described in https://issues.redhat.com/browse/KEYCLOAK-18368. But at the same time, we should make sure that offline tokens from previous version must be possible to refresh. So that's the story of why it was originally added :-)


Now some thoughts for this PR:

IMO the ideal is, if this fallback to 0 (or timestamp) is effectively not needed anywhere.  It looks that the only case how the USER_SESSION_STARTED_AT_NOTE note won't be present is (again) the offline sessions migrated from previous version?

AFAIK we don't need to care about new store as we don't support migrating of offline sessions created with the new store in previous versions. Is it correct @martin-kanis ? So we likely need to care just about old store and the offline sessions coming from `UserSessionPersister`.

So as long as all client sessions returned by `JpaUserSessionPersisterProvider` have this note on them, we should be good. If we always have corresponding UserSession (which AFAIK we have), we can just use `userSession.getStarted()` to fill that note before saving those persisted sessions in infinispan?

Also I wonder if `AuthenticatedClientSessionModel.getUserSessionStarted()` could be updated like this in this PR?

```
    default int getUserSessionStarted() {
        String started = getNote(USER_SESSION_STARTED_AT_NOTE);
        return started == null ? getUserSession().getStarted() : Integer.parseInt(started);
    }
```
In other words, fallback to look at `userSession.getStarted()` instead of using hardcoded value of 0 (or timestamp). Ideal is if AuthenticatedClientSessionEntity also has access to userSession, so it can look into it, but I suppose this would require some bigger refactoring, which we don't want to do in this PR?


In shortcut: In `MigrationTest` we have scenario for testing refresh of the offline-token from previous version. So it seems that this PR should be tested with `MigrationTest` to make sure it works. AFAIK `MigrationTest` is not yet in GH actions, so this may need to be done manually.

rmartinc replied:
I have tested this with the `MigrationTest` and it works (`mariadb` only, but I think that's enough). I have also done two little changes:

* The change proposed by @mposolda in `AuthenticatedClientSessionModel.getUserSessionStarted`. 
* And I also modified [these lines](https://github.com/keycloak/keycloak/pull/17525/files#diff-fd8b1c47da5ec7ee42d5450d3f7b73cdeef9c08af26b825c9245e3e6052c5cdfR1090) commented by @martin-kanis. He is right and if 0 the previous check for `isIssuedBeforeSessionStart` fails.

For the previous comments, I think that returning 0 in `AuthenticatedClientSessionModel` interface is necessary. Not changing anything there in this PR for the moment.

WDYT now?

martin-kanis replied:
@mposolda Thanks for the explanation. Also the proposed changes make sense to me. So @rmartinc LGTM

martin-kanis commented on model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/entities/AuthenticatedClientSessionEntity.java:
```diff
@@ -83,6 +85,28 @@ public void setTimestamp(int timestamp) {
+        return started == null ? timestamp : Integer.parseInt(started);
+    }
+
+    public int getStarted() {
```
This defaults to `timestamp` while `AuthenticatedClientSessionModel.getStarted` defaults to 0 if `STARTED_AT_NOTE` is not set. Is it intentional?

rmartinc replied:
Yep, the same.

martin-kanis reviewed (approved):
@rmartinc Thank you for the PR and all the changes! LGTM

mposolda reviewed (approved):
@rmartinc @martin-kanis Thanks for the great work and for the review!
